### PR TITLE
New post: Fine-tuning Stable Diffusion Models on Intel CPUs

### PR DESCRIPTION
**Pull request by juliensimon**

### CODE REVIEW

**Comment by juliensimon** [comment]
I'll keep tweaking and fixing, but please feel free to send feedback now.

**Review by jeffboudier — APPROVED** [review]
Great post, thanks for the detailed walkthrough Julien!

**Review by VioletteLepercq — APPROVED** [review]

File: stable-diffusion-finetuning-intel.md [hunk 1]
```diff
@@ -0,0 +1,251 @@
+---
+title: "Fine-tuning Stable Diffusion models on Intel CPUs"
+thumbnail: /blog/assets/stable-diffusion-finetuning-intel/01.png
+authors:
+- user: juliensimon
+---
```
Comment (pcuenca) [inline stable-diffusion-finetuning-intel.md:6]: Don't forget to update `_blog.yml` when you are done :) Also, a thumbnail needs to be added to `assets`.

File: stable-diffusion-finetuning-intel.md [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+192.168.20.2
+192.168.21.2
+192.168.22.2
+192.168.23.2
+EOF
+```
+
+Distributed training requires password-less `ssh` between the primary and other nodes. Here's a good [article](https://www.redhat.com/sysadmin/passwordless-ssh) on how to do this if you're unfamiliar with the process.
+
+Next, we create a new environment on each node and install the software dependencies. We notably install two Intel libraries: [oneCCL](https://github.com/oneapi-src/oneCCL), to manage distributed communication and the [Intel Extension for PyTorch](https://github.com/intel/intel-extension-for-pytorch) (IPEX) to leverage the hardware acceleration features present in Sapphire Rapids. We also add `gpertools` to install `libtcmalloc,` a high-performance memory allocation library.
```
Comment (pcuenca) [inline stable-diffusion-finetuning-intel.md:60]:
```suggestion
Next, we create a new environment on each node and install the software dependencies. We notably install two Intel libraries: [oneCCL](https://github.com/oneapi-src/oneCCL), to manage distributed communication and the [Intel Extension for PyTorch](https://github.com/intel/intel-extension-for-pytorch) (IPEX) to leverage the hardware acceleration features present in Sapphire Rapids. We also add `gperftools` to install `libtcmalloc,` a high-performance memory allocation library. 
```

File: stable-diffusion-finetuning-intel.md [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+conda create -n diffuser python==3.9
+conda activate diffuser
+pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cpu
+pip3 install transformers accelerate==0.19.0
+pip3 install oneccl_bind_pt -f https://developer.intel.com/ipex-whl-stable-cpu
+pip3 install intel_extension_for_pytorch
+conda install gperftools -c conda-forge -y
+```
+
+Next, we clone the [diffusers](https://github.com/huggingface/diffusers/) repository on each node and install it from the source.
```
Comment (pcuenca) [inline stable-diffusion-finetuning-intel.md:72]:
```suggestion
Next, we clone the [diffusers](https://github.com/huggingface/diffusers/) repository on each node and install it from source.
```

File: stable-diffusion-finetuning-intel.md [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
++    import intel_extension_for_pytorch as ipex
++    unet = ipex.optimize(unet, dtype=weight_dtype)
++    vae = ipex.optimize(vae, dtype=weight_dtype)
++
+     # We need to recalculate our total training steps as the size of the training dataloader may have changed.
+     num_update_steps_per_epoch = math.ceil(len(train_dataloader) / args.gradient_accumulation_steps)
+     if overrode_max_train_steps:
+```
+
+The last step is downloading the [training images](https://huggingface.co/sd-concepts-library/dicoo). Ideally, we'd use a shared NFS folder, but for the sake of simplicity, we'll download the images on each node. Please ensure they're in the same directory on all nodes (`/home/devcloud/dicoo`).
```
Comment (pcuenca) [inline stable-diffusion-finetuning-intel.md:101]:
```suggestion
The last step is downloading the [training images](https://huggingface.co/sd-concepts-library/dicoo). For the sake of simplicity, we'll just download the images on each node. Please ensure they're in the same directory on all nodes (`/home/devcloud/dicoo`).
```

(It's just a few images)

File: stable-diffusion-finetuning-intel.md [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+wget https://huggingface.co/sd-concepts-library/dicoo/resolve/main/concept_images/4.jpeg
+```
+
+Here are the images:
+
+<img src="assets/stable-diffusion-finetuning-intel/0.jpeg" height="256">
+<img src="assets/stable-diffusion-finetuning-intel/1.jpeg" height="256">
+<img src="assets/stable-diffusion-finetuning-intel/2.jpeg" height="256">
+<img src="assets/stable-diffusion-finetuning-intel/3.jpeg" height="256">
+<img src="assets/stable-diffusion-finetuning-intel/4.jpeg" height="256">
```
Comment (pcuenca) [inline stable-diffusion-finetuning-intel.md:119]:
```suggestion
<img src="https://huggingface.co/sd-concepts-library/dicoo/resolve/main/concept_images/0.jpeg" height="256">
<img src="https://huggingface.co/sd-concepts-library/dicoo/resolve/main/concept_images/1.jpeg" height="256">
<img src="https://huggingface.co/sd-concepts-library/dicoo/resolve/main/concept_images/2.jpeg" height="256">
<img src="https://huggingface.co/sd-concepts-library/dicoo/resolve/main/concept_images/3.jpeg" height="256">
<img src="https://huggingface.co/sd-concepts-library/dicoo/resolve/main/concept_images/4.jpeg" height="256">
```

Are they not the same? Otherwise, we usually upload images (except the thumbnail) to this repo https://huggingface.co/datasets/huggingface/documentation-images/tree/main/blog to keep the blog repo lean (but I think we can just remove them in this case, and use the original URLs).

File: stable-diffusion-finetuning-intel.md [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+The system setup is now complete. Let's configure the training job.
+
+## Configuring the fine-tuning job
+
+The [Accelerate](https://huggingface.co/docs/accelerate/index) library makes it very easy to run distributed training. We need to run it on each node and answer simple questions.
+
+Here's a screenshot for the primary node. On the other nodes, you need to set the rank to 1, 2, and 3. All other answers are identical.
+
+<kbd>
+  <img src="assets/stable-diffusion-finetuning-intel/screen01.png">
```
Comment (pcuenca) [inline stable-diffusion-finetuning-intel.md:131]: Ideally to be moved to https://huggingface.co/datasets/huggingface/documentation-images/tree/main/blog, creating a folder for this post.

File: stable-diffusion-finetuning-intel.md [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+</kbd>
+
+Finally, we need to set the environment on the primary node. It will be propagated to other nodes as the fine-tuning job starts. The first line sets the name of the network interface connected to the local network where all nodes run. You may need to adapt this: `ifconfig` will show you the appropriate information.
```
Comment (pcuenca) [inline stable-diffusion-finetuning-intel.md:134]:
```suggestion
Finally, we need to set the environment on the primary node. It will be propagated to other nodes as the fine-tuning job starts. The first line sets the name of the network interface connected to the local network where all nodes run. You may need to adapt this using`ifconfig` to get the appropriate information.
```

File: stable-diffusion-finetuning-intel.md [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+--max_train_steps=3000 --learning_rate=2.0e-03 --scale_lr --lr_scheduler="constant"     \
+--lr_warmup_steps=0 --output_dir=./textual_inversion_output --mixed_precision bf16      \
+--save_as_full_pipeline
+```
+
+If training starts successfully, stop it and move to the next node. If training starts successfully on all nodes, return to the primary node and double-check the node file, the environment, and the `mpirun` command. Don't worry; you'll find the problem :)
+
+## Generating images with the fine-tuned model
+
+After an hour or so of training, the model is saved locally. We could load it with a vanilla `diffusers` pipeline and predict. Instead, let's use [Optimum Intel and OpenVINO](https://huggingface.co/docs/optimum/intel/inference) to optimize the model. As discussed in a [previous post](https://huggingface.co/blog/intel-sapphire-rapids-inference), this lets you generate an image on a single CPU in less than 5 seconds!
```
Comment (pcuenca) [inline stable-diffusion-finetuning-intel.md:194]: 👍

File: stable-diffusion-finetuning-intel.md [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+
+Like Transformer models, you can fine-tune Diffusion models to help them generate content that matches your business needs. Initially, fine-tuning was only possible on GPU infrastructure, but things are changing! A few months ago, Intel launched the fourth generation of Xeon CPUs, code-named Sapphire Rapids. Sapphire Rapids introduces the Intel Advanced Matrix Extensions (AMX), a new hardware accelerator for deep learning workloads. We've already demonstrated the benefits of AMX in several blog posts: [fine-tuning NLP Transformers](https://huggingface.co/blog/intel-sapphire-rapids), [inference with NLP Transformers](https://huggingface.co/blog/intel-sapphire-rapids-inference), and [inference with Stable Diffusion models](https://huggingface.co/blog/intel-sapphire-rapids-inference). 
+
+This post will show you how to fine-tune a Stable Diffusion model on a cluster on Intel Sapphire Rapids CPUs. We will use [textual inversion](https://huggingface.co/docs/diffusers/training/text_inversion), a technique that only requires a small number of example images. We'll use only five!
```
Comment (pcuenca) [inline stable-diffusion-finetuning-intel.md:18]: Perhaps we could mention that other methods such as Dreambooth are also possible, even if we don't cover them in this guide. Just mentioning it because Dreambooth is very very popular, and it supports LoRA for faster training.

File: stable-diffusion-finetuning-intel.md [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+--resolution=512  --train_batch_size=1  --seed=7  --gradient_accumulation_steps=1       \
+--max_train_steps=3000 --learning_rate=2.0e-03 --scale_lr --lr_scheduler="constant"     \
+--lr_warmup_steps=0 --output_dir=./textual_inversion_output --mixed_precision bf16      \
+--save_as_full_pipeline
+```
+
+Here's a screenshot of the busy cluster:
+
+<kbd>
+  <img src="assets/stable-diffusion-finetuning-intel/screen02.png">
```
Comment (pcuenca) [inline stable-diffusion-finetuning-intel.md:171]: Ideally to be moved as well.

File: stable-diffusion-finetuning-intel.md [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+images = ov_pipe(prompt).images
+print(images)
+for idx,img in enumerate(images):
+    img.save(f"image{idx}.png")
+```
+
+Here's a rather lovely example. It is impressive that the model only needed five images to learn that dicoos have glasses!
+
+<kbd>
+  <img src="assets/stable-diffusion-finetuning-intel/dicoo_image.png">
```
Comment (pcuenca) [inline stable-diffusion-finetuning-intel.md:230]: Same.

**Review by pcuenca — COMMENTED** [review]
Very cool!

File: stable-diffusion-finetuning-intel.md (hunk 4 above)
Comment (juliensimon) [reply stable-diffusion-finetuning-intel.md:101]: I see your point, but in general, NFS would be a better option :)

File: stable-diffusion-finetuning-intel.md (hunk 5 above)
Comment (juliensimon) [reply stable-diffusion-finetuning-intel.md:119]: good idea, why didn't I think about this?

**Comment by osanseviero** [comment]
As mentioned by @pcuenca, the assets (except the thumbnail) should be on the Hub repository to keep this repository lighter and easier to work with.
